### ISSUE TITLE
CamelCaseParameterName and CamelCaseVariableName doesn't allow underscore name

### ISSUE THREAD

**Issue by a-menshchikov** [phpmd/phpmd#740]
In recent years, it has become the best practice to name unused variable, that cannot be removed from code (caught exception that doesn't use, for example) with name `$_`.
The same practice has existed for many years in JavaScript, Golang and other communities.

I think that CamelCaseParameterName and CamelCaseVariableName rules should not raise violation on this parameter/variable name.

**Comment by kylekatarnls** [comment]
This should be a dedicated rule I guess, because it means it's an exception for CamelCaseParameterName, CamelCaseVariableName and UnusedVariableRule, then it would in fact be mandatory for this variable not to be used, so we should raise an error if `$_` is read or write after its initialization.

**Comment by tuc0w** [comment]
If CamelCaseParameterName and CamelCaseVariableName would implement the `allow-underscore` and `allow-underscore-test` properties wouldn't it be solved?

Since CamelCasePropertyName and CamelCaseMethodName already allow those properties it wouldn't be an exception, right?

**Comment by kylekatarnls** [comment]
Indeed, but it's not the same purpose and result. The "good" rule would be to only accept `$_` (variable name exactly equal to "_") when assigned and not read. Using an option like `allow_underscore` would allow to have any variable name starting with an "_" (so also `$_abc` and it would allow it in any context no mater it's an assignation or not, and even if it's read later. So yes, it would chill your console output but it would allow way more (and probably unwanted) variables names and situations.

**Comment by kylekatarnls** [comment]
I'm still open to for either the allow_underscore solution (as nothing prevent us to implement this option in every naming rules). It would even be great for consistency, or the full implementation of the `$_` unused variable pattern as its own rule. PRs for any of them would be welcome.

**Comment by tuc0w** [comment]
> Indeed, but it's not the same purpose and result. The "good" rule would be to only accept `$_` (variable name exactly equal to `"_"`) when assigned and not read. Using an option like `allow_underscore` would allow to have any variable name starting with an `"_"` (so also `$_abc` and it would allow it in any context no mater it's an assignation or not, and even if it's read later. So yes, it would chill your console output but it would allow way more (and probably unwanted) variables names and situations.

You're right, that's the part I forgot because I encountered the `$_abc` problem today and thought this issue would be the solution. Of course a dedicated rule would be the clean way to implement the `$_` solution.
